components/router/history.js

```diff
@@ -3,44 +3,82 @@
  *  Licensed under the MIT License. See License.txt in the project root for license information.
  *--------------------------------------------------------------------------------------------*/
 
-class BrowserHistory {
+class History {
     constructor() {
-        this.listeners = [];
-        this.currentURL = '';
-        this.currentRouteId = 0;
+        this.currentRouteId = -1;
+        this.onHistoryChange = null;
+        this.onBeforeNavigation = null;
+        this._onPopStateBound = this._onPopState.bind(this);
     }
 
-    /**
-     * Adds a callback to the listeners array.
-     * @param {function} callback - the function that will be called on dispatch.
-    */
-    listen(callback) {
-        this.listeners.push(callback);
+    get currentState() {
+        if (!window.history || !window.history.state || !window.history.state.current) return '';
+        return window.history.state.current;
     }
 
     get nextRouteId() {
         return this.currentRouteId + 1;
     }
 
     /**
-     * Calls all functions added to the listeners array.
-    */
-    dispatch() {
-        for (let listener of this.listeners) {
-            listener(this.currentURL);
+     * Will call the history change callback if there is one
+     */
+    _dispatchHistoryChange() {
+        if (!this.onBeforeNavigation) {
+            if (this.onHistoryChange) this.onHistoryChange();

```

Comment:
Why did we remove the listeners. Now we can only listen to one event which is onHistoryChange.

Reply:
Because there was just a single listener used in the router js. I didn't see any use case in the previous implementation to have an array with listeners that listeners are called when the history has changed. So I decided to transform it to a more readable and a single 'event/callback' - `onHistoryChange`. If you know any good reason to have that listeners logic I will return back the code.